cas-client-core/src/main/java/org/jasig/cas/client/authentication/DefaultGatewayResolverImpl.java

```diff
@@ -33,7 +33,6 @@ public boolean hasGatewayedAlready(final HttpServletRequest request, final Strin
         }
 
         final boolean result = session.getAttribute(CONST_CAS_GATEWAY) != null;
-        session.removeAttribute(CONST_CAS_GATEWAY);
```

Comment:
wondering why is this line removed, this change seems to make an opened tab not able auto login after other tab login and refresh the opened one.

Reply:
also I can not find #177

Reply:
> wondering why is this line removed, this change seems to make an opened tab not able auto login after other tab login and refresh the opened one.

You might want to check HOW the new tab is being opened. 
If the tab is opened following an action external to the site itself, (a mail client, an URL in another webiste, an URL in a desktop app, etc), the SameSite cookie policy configured for the JavaEE session may prevent the session from being used in the new tab. But this has nothing to do with CAS or your application, this is security settings. 
Some browser force the SameSite settings if it was not specified by the server when sending the cookie (see https://developers.google.com/search/blog/2020/01/get-ready-for-new-samesitenone-secure).
You can try to configure it the way you want, for example in Tomcat : 
https://tomcat.apache.org/tomcat-9.0-doc/config/cookie-processor.html

> also I can not find #177 everywhere

It looks like issues in this repository have been made private, only pull request are still accessible... maybe ask a maintainer.
The issue was available at https://github.com/apereo/java-cas-client/issues/177

Reply:
Thank you much for quick reply !

While I still have some problems from my project after some experiments and thoughts.

Let's say I have two site *A* and *B*, both use java-cas-client 3.6.4 as its lib and enable `gateway` feature for `AuthenticationFilter`, and let's assume *C* is the CAS server.

I do the following steps with the three web site:

0. make sure `A` and `B` are both logged out
1. open site A by entering url a.com
2. open a new tab, open site B by entering url b.com

now both A and B servers has their own session created and attribute `CONST_CAS_GATEWAY` set to "yes" because

https://github.com/apereo/java-cas-client/blob/146a7c67a4ae1aa8adcf4b2bf41ffea286787401/cas-client-core/src/main/java/org/apereo/cas/client/authentication/AuthenticationFilter.java#L201

https://github.com/apereo/java-cas-client/blob/146a7c67a4ae1aa8adcf4b2bf41ffea286787401/cas-client-core/src/main/java/org/apereo/cas/client/authentication/DefaultGatewayResolverImpl.java#L41-L45

3. now I switch to the opened site *B* b.com, and login as user *U*.
4. go back to opened site *A*, because *A* has already have session with CONST_CAS_GATEWAY set to "yes" at step 1, when I refresh the page or open a new tab with url a.com, the user *U* signed in at site *B* will never be signed automatically in site *A*, since the rediection to C (c.com) will not happen. because `wasGatewayed` is always true, and filter goes to next one at line 192

https://github.com/apereo/java-cas-client/blob/146a7c67a4ae1aa8adcf4b2bf41ffea286787401/cas-client-core/src/main/java/org/apereo/cas/client/authentication/AuthenticationFilter.java#L189-L200

Reply:
seems `this.gatewayStorage.hasGatewayedAlready(request, serviceUrl)` alway add `CONST_CAS_GATEWAY` back to session, no matter this line exists or not